DNS Controller throws panic on multiple entries deletion

**How to categorize this issue?**
<!--
Please select area, kind, and priority for this issue. This helps the community categorizing it.
Replace below TODOs or exchange the existing identifiers with those that fit best in your opinion.
If multiple identifiers make sense you can also state the commands multiple times, e.g.
  /area control-plane
  /area auto-scaling
  ...

"/area" identifiers:     audit-logging|auto-scaling|backup|certification|control-plane-migration|control-plane|cost|delivery|dev-productivity|disaster-recovery|documentation|high-availability|logging|metering|monitoring|networking|open-source|operations|ops-productivity|os|performance|quality|robustness|scalability|security|storage|testing|usability|user-management
"/kind" identifiers:     api-change|bug|cleanup|discussion|enhancement|epic|impediment|poc|post-mortem|question|regression|task|technical-debt|test
"/priority" identifiers: normal|critical|blocker
-->
/area robustness
/kind bug
/priority normal

**What happened**:
The DNS Controller enters in a CrashLoopBackOff during multiple entries deletion.
Impacts the whole Seed cluster and users cannot provision DNS using the controller.

```
panic: runtime error: slice bounds out of range [3:1] [recovered]
	panic: runtime error: slice bounds out of range [3:1]

goroutine 680 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1a31020, 0xc01750d900)
	/usr/local/go/src/runtime/panic.go:969 +0x166
github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/common.(*StateHandler).EnsureEntries(0xc004b990e0, 0x2cfdb00, 0x0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/common/state.go:153 +0x440
github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/common.(*StateHandler).Refresh(0xc004b990e0, 0xc005fde240, 0x1df3f60, 0xc0001d0390, 0xc001f885c0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/common/state.go:139 +0x8d
github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/common.NewStateHandler(0x1ddee20, 0xc000372000, 0xc000c6ad00, 0xc00340a340, 0xc003938501, 0x0, 0x0, 0x0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/common/state.go:78 +0x37f
github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/lifecycle.(*actuator).deleteSeedResources(0xc000994420, 0x1ddee20, 0xc000372000, 0xc00340a340, 0x0, 0x0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/lifecycle/actuator.go:291 +0x406
github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/lifecycle.(*actuator).Delete(0xc000994420, 0x1ddee20, 0xc000372000, 0xc00340a340, 0x1b3f93e, 0x6)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/pkg/controller/lifecycle/actuator.go:192 +0x4d
github.com/gardener/gardener/extensions/pkg/controller/extension.(*reconciler).delete(0xc000a86180, 0x1ddee20, 0xc000372000, 0xc00340a340, 0x2cff020, 0x1a, 0xc000a29180, 0x0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/github.com/gardener/gardener/extensions/pkg/controller/extension/reconciler.go:239 +0x1da
github.com/gardener/gardener/extensions/pkg/controller/extension.(*reconciler).Reconcile(0xc000a86180, 0xc000b10ea0, 0x1a, 0xc000b10e80, 0x11, 0x0, 0x11, 0x1da4a20, 0xc00340a000)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/github.com/gardener/gardener/extensions/pkg/controller/extension/reconciler.go:193 +0x82a
github.com/gardener/gardener/extensions/pkg/controller.(*operationAnnotationWrapper).Reconcile(0xc0008d0fc0, 0xc000b10ea0, 0x1a, 0xc000b10e80, 0x11, 0x0, 0xbfd97860efdf10d2, 0xc0007f2123, 0xc00022efc8)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/github.com/gardener/gardener/extensions/pkg/controller/reconciler.go:90 +0x283
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc00015c000, 0x198a4e0, 0xc0021ca900, 0xc0001f9e00)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:245 +0x161
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc00015c000, 0xc0001f9f00)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:221 +0xae
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker(0xc00015c000)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:200 +0x2b
k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000427cb0)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x5f
k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000427cb0, 0x3b9aca00, 0x0, 0x1, 0xc00073e360)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
k8s.io/apimachinery/pkg/util/wait.Until(0xc000427cb0, 0x3b9aca00, 0xc00073e360)
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x4d
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func1
	/go/src/github.com/gardener/gardener-extension-shoot-dns-service/vendor/sigs.k8s.io/controller-runtime/pkg/internal/controller/controller.go:182 +0x526
```

**What you expected to happen**:
To handle the deletion successfully 

**How to reproduce it (as minimally and precisely as possible)**:

**Anything else we need to know?**:

**Environment**:
aws

- Gardener version (if relevant):n- Extension version:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- Others:

--- a/pkg/controller/common/state.go
+++ b/pkg/controller/common/state.go
@@ -143,14 +143,15 @@ func (s *StateHandler) EnsureEntries(entries []dnsapi.DNSEntry) bool {
 	mod := false
 	names := sets.String{}
 	for _, entry := range entries {
-		mod = s.EnsureEntryFor(&entry)
+		mod = s.EnsureEntryFor(&entry) || mod
 		names.Insert(entry.Name)
 	}
 	if len(entries) != len(s.state.Entries) {
-		for i, e := range s.state.Entries {
+		for i := len(s.state.Entries) - 1; i >= 0; i-- {
+			e := s.state.Entries[i]
 			if !names.Has(e.Name) {
-				mod = true
 				s.state.Entries = append(s.state.Entries[:i], s.state.Entries[i+1:]...)
+				mod = true
 			}
 		}
 	}

--- a/pkg/controller/common/state_test.go
+++ b/pkg/controller/common/state_test.go
@@ -1,0 +1,52 @@
+/*
+ * Copyright 2020 SAP SE or an SAP affiliate company. All rights reserved. This file is licensed under the Apache Software License, v. 2 except as noted otherwise in the LICENSE file
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ *
+ */
+
+package common
+
+import (
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	v1alpha1 "github.com/gardener/external-dns-management/pkg/apis/dns/v1alpha1"
+	api "github.com/gardener/gardener-extension-shoot-dns-service/pkg/apis"
+)
+
+var _ = Describe("StateHandler", func() {
+	DescribeTable("#EnsureEntries",
+		func(expectedMod bool, newEntryNames ...string) {
+			handler := &StateHandler{state: &api.DNSState{}}
+			handler.state.Entries = []*api.DNSEntry{
+				{Name: "entry1", Spec: &v1alpha1.DNSEntrySpec{}},
+				{Name: "entry2", Spec: &v1alpha1.DNSEntrySpec{}},
+			}
+			var newEntries []v1alpha1.DNSEntry
+			for _, name := range newEntryNames {
+				newEntries = append(newEntries, v1alpha1.DNSEntry{ObjectMeta: metav1.ObjectMeta{Name: name}})
+			}
+			mod := handler.EnsureEntries(newEntries)
+			Expect(mod).To(Equal(expectedMod))
+			Expect(len(handler.state.Entries)).To(Equal(len(newEntries)))
+		},
+		Entry("unchanged", false, "entry1", "entry2"),
+		Entry("unchanged2", false, "entry2", "entry1"),
+		Entry("add/remove", true, "entry1", "entry3"),
+		Entry("delete one", true, "entry1"),
+		Entry("replace all", true, "entry3", "entry4"),
+		Entry("none", true),
+	)
+})

